tests/test_pin.py

```diff
@@ -1,76 +1,162 @@
+from unittest import TestCase
 
 from ddtrace import Pin
-from nose.tools import eq_
-
-
-def test_pin():
-    class A(object):
-        pass
-
-    a = A()
-    pin = Pin(service="abc")
-    pin.onto(a)
-
-    got = Pin.get_from(a)
-    assert pin.service == got.service
-    assert pin is got
-
-def test_cant_pin():
-
-    class Thing(object):
-        __slots__ = ['t']
-
-    t = Thing()
-    t.t = 1
-
-    Pin(service="a").onto(t)
-
-def test_cant_modify():
-    p = Pin(service="abc")
-    try:
-        p.service = "other"
-    except AttributeError:
-        pass
-
-def test_copy():
-    p1 = Pin(service="a", app="app_type", tags={"a":"b"})
-    p2 = p1.clone(service="b")
-    assert p1.service == "a"
-    assert p2.service == "b"
-    assert p1.app == "app_type"
-    assert p2.app == "app_type"
-    eq_(p1.tags, p2.tags)
-    assert not (p1.tags is p2.tags)
-    assert p1.tracer is p2.tracer
-
-def test_none():
-    assert None is Pin.get_from(None)
-
-def test_repr():
-    p = Pin(service="abc")
-    assert p.service == "abc"
-    assert 'abc' in str(p)
-
-def test_override():
-    class A(object):
-        pass
-
-    Pin(service="foo", app="blah").onto(A)
-    a = A()
-    Pin.override(a, app="bar")
-    eq_(Pin.get_from(a).app, "bar")
-    eq_(Pin.get_from(a).service, "foo")
-
-    b = A()
-    eq_(Pin.get_from(b).service, "foo")
-    eq_(Pin.get_from(b).app, "blah")
-
-
-def test_overide_missing():
-    class A():
-        pass
-
-    a = A()
-    assert not Pin.get_from(a)
-    Pin.override(a, service="foo")
-    assert Pin.get_from(a).service == "foo"
+from nose.tools import eq_, ok_, assert_raises
+
+
+class PinTestCase(TestCase):
+    """TestCase for the `Pin` object that is used when an object is wrapped
+    with our tracing functionalities.
+    """
+    def setUp(self):
+        # define a simple class object
+        class Obj(object):
+            pass
+
+        self.Obj = Obj
+
+    def test_pin(self):
+        # ensure a Pin can be attached to an instance;
+        # the first time the `pin` is used via `get_from`, it gets cloned
+        obj = self.Obj()
+        pin = Pin(service='metrics')
+        pin.onto(obj)
+
+        got = Pin.get_from(obj)
+        eq_(got.service, pin.service)
+        ok_(got is not pin)
```

Comment:
This test before was `ok_(got is pin)`. We now clone the pin once, when `get_from` is used to safe-protect mutable structure like `config` and `tags`. Not doing that means introducing a bug where an instance can update global configurations (and we don't want that).

Reply:
Fixed using another approach.